driver/registry_default.go

```diff
@@ -586,71 +586,77 @@ func (m *RegistryDefault) Init(ctx context.Context, ctxer contextx.Contextualize
+		m.WithContextualizer(ctxer)
+
+		pool, idlePool, connMaxLifetime, connMaxIdleTime, cleanedDSN := sqlcon.ParseConnectionOptions(m.l, m.Config().DSN(ctx))
+		m.Logger().
+			WithField("pool", pool).
+			WithField("idlePool", idlePool).
+			WithField("connMaxLifetime", connMaxLifetime).
+			Debug("Connecting to SQL Database")
+		c, err := pop.NewConnection(&pop.ConnectionDetails{
+			URL:                       sqlcon.FinalizeDSN(m.l, cleanedDSN),
+			IdlePool:                  idlePool,
+			ConnMaxLifetime:           connMaxLifetime,
+			ConnMaxIdleTime:           connMaxIdleTime,
+			Pool:                      pool,
+			UseInstrumentedDriver:     m.Tracer(ctx).IsLoaded(),
+			InstrumentedDriverOptions: instrumentedDriverOpts,
+		})
+		if err != nil {
+			m.Logger().WithError(err).Warnf("Unable to connect to database, retrying.")
+			return errors.WithStack(err)
+		}
+		if err := c.Open(); err != nil {
+			m.Logger().WithError(err).Warnf("Unable to open database, retrying.")
+			return errors.WithStack(err)
+		}
+		p, err := sql.NewPersister(ctx, m, c)
+		if err != nil {
+			m.Logger().WithError(err).Warnf("Unable to initialize persister, retrying.")
+			return err
+		}
 
-			if err := p.Ping(); err != nil {
-				m.Logger().WithError(err).Warnf("Unable to ping database, retrying.")
+		if err := p.Ping(); err != nil {
+			m.Logger().WithError(err).Warnf("Unable to ping database, retrying.")
+			return err
+		}
+
+		// if dsn is memory we have to run the migrations on every start
+		if dbal.IsMemorySQLite(m.Config().DSN(ctx)) || m.Config().DSN(ctx) == "memory" {
+			m.Logger().Infoln("Ory Kratos is running migrations on every startup as DSN is memory. This means your data is lost when Kratos terminates.")
+			if err := p.MigrateUp(ctx); err != nil {
+				m.Logger().WithError(err).Warnf("Unable to run migrations, retrying.")
 				return err
 			}
+		}
 
-			// if dsn is memory we have to run the migrations on every start
-			if dbal.IsMemorySQLite(m.Config().DSN(ctx)) || m.Config().DSN(ctx) == "memory" {
-				m.Logger().Infoln("Ory Kratos is running migrations on every startup as DSN is memory. This means your data is lost when Kratos terminates.")
-				if err := p.MigrateUp(ctx); err != nil {
-					m.Logger().WithError(err).Warnf("Unable to run migrations, retrying.")
-					return err
-				}
-			}
+		if o.skipNetworkInit {
+			m.persister = p
+			return nil
+		}
 
-			if o.skipNetworkInit {
-				m.persister = p
-				return nil
-			}
+		net, err := p.DetermineNetwork(ctx)
+		if err != nil {
+			m.Logger().WithError(err).Warnf("Unable to determine network, retrying.")
+			return err
+		}
 
-			net, err := p.DetermineNetwork(ctx)
-			if err != nil {
-				m.Logger().WithError(err).Warnf("Unable to determine network, retrying.")
-				return err
-			}
+		m.persister = p.WithNetworkID(net.ID)
+		return nil
+	}, bc)
 
-			m.persister = p.WithNetworkID(net.ID)
-			return nil
-		}, bc),
-	)
+	if err != nil {
+		return err
+	}
+
+	if o.inspect != nil {
```

Comment:
Hm not sure what the use-case is, but I see two code-paths that `return nil` before this is reached. Just FYI.

Reply:
@alnr please double check this

Reply:
You mean on lines 637 and 647? Those returns are returning from the closure passed to `backoff.Retry`. We return from `Init` only if `backoff.Retry` returns non-nil. So this codepath is always reached in the happy path.